File: node/be/api/steps/journal.steps.js
'use strict';

const { When, Then } = require('@cucumber/cucumber');
const { MiniBooks, ApiUnreachable } = require('../venues/books');

/**
 * Steps for features/be-mini-books-journal.feature. The generic "refused with
 * code/status" steps live here and are shared across every @books @api tier.
 */

const books = new MiniBooks();

async function act(world, fn) {
  if (world.sourceError) return undefined;
  try { return await fn(); } catch (err) { if (err instanceof ApiUnreachable) { world.sourceError = err.message; return undefined; } throw err; }
}
function check(world, description, fn) {
  if (world.sourceError) { world.unobservable(description, 'the source could not be reached -- ' + world.sourceError); return; }
  const r = fn(); world.check(description, r.passed, r.detail);
}
const j = (v) => JSON.stringify(v);

async function postLines(world, lines, token, { date = books.bizDay, key } = {}) {
  const body = { entry_date: date, lines };
  if (key) body.idempotency_key = key;
  await act(world, async () => { world.api = await books.post('/journal-entries', body, token ? { token } : {}); if (world.api.status === 201 || world.api.status === 200) world.entry = world.api.body; });
}
const twoLine = (dr, cr, amt) => [{ account_id: dr, side: 'debit', amount_cents: amt }, { account_id: cr, side: 'credit', amount_cents: amt }];

When('the accountant posts Dr account {int} Cr account {int} for {int}', async function (dr, cr, amt) { await postLines(this, twoLine(dr, cr, amt), this.token); });
When('the accountant posts Dr account {int} Cr account {int} for {int} dated {word}', async function (dr, cr, amt, date) { await postLines(this, twoLine(dr, cr, amt), this.token, { date }); });
When('the accountant posts a split entry debiting account {int} by {int} and account {int} by {int} crediting account {int} by {int}', async function (d1, a1, d2, a2, cr, ac) { await postLines(this, [{ account_id: d1, side: 'debit', amount_cents: a1 }, { account_id: d2, side: 'debit', amount_cents: a2 }, { account_id: cr, side: 'credit', amount_cents: ac }], this.token); });
When('the accountant posts an unbalanced entry of Dr {int} Cr {int}', async function (dr, cr) { await postLines(this, [{ account_id: 1, side: 'debit', amount_cents: dr }, { account_id: 9, side: 'credit', amount_cents: cr }], this.token); });
When('the accountant posts a single-line entry', async function () { await postLines(this, [{ account_id: 1, side: 'debit', amount_cents: 1000 }], this.token); });
When('an anonymous caller posts Dr account {int} Cr account {int} for {int}', async function (dr, cr, amt) { await postLines(this, twoLine(dr, cr, amt), null); });
When('a forged token posts Dr account {int} Cr account {int} for {int}', async function (dr, cr, amt) { await postLines(this, twoLine(dr, cr, amt), 'usr_forged000000000000000000'); });
When('the accountant posts Dr account {int} Cr account {int} for {int} with key {string}', async function (dr, cr, amt, key) { await postLines(this, twoLine(dr, cr, amt), this.token, { key }); this.firstEntry = this.api; this.lastLines = twoLine(dr, cr, amt); });
When('the accountant posts again with key {string}', async function (key) { await act(this, async () => { this.api = await books.post('/journal-entries', { entry_date: books.bizDay, lines: this.lastLines, idempotency_key: key }, { token: this.token }); this.secondEntry = this.api; }); });

When('the accountant reads the entry', async function () { await act(this, async () => { this.api = await books.get('/journal-entries/' + this.entry.id, { token: this.token }); }); });
When('the accountant lists the entries', async function () { await act(this, async () => { this.api = await books.get('/journal-entries', { token: this.token }); }); });

Then('the entry is posted', function () {
  check(this, 'entry posted', () => ({ passed: !!this.api && this.api.status === 201, detail: this.api ? this.api.status + ' ' + j(this.api.body) : 'no response' }));
});
Then('the entry is refused', function () {
  check(this, 'entry refused', () => ({ passed: !!this.api && this.api.status >= 400, detail: this.api ? this.api.status + ' ' + j(this.api.body) : 'no response' }));
});
Then('the entry response balances', function () {
  check(this, 'entry response balances', () => { const e = this.entry; if (!e || !e.lines) return { passed: false, detail: this.api ? this.api.status + ' ' + j(this.api.body) : 'no entry' }; const d = e.lines.filter((l) => l.side === 'debit').reduce((a, l) => a + l.amount_cents, 0); const c = e.lines.filter((l) => l.side === 'credit').reduce((a, l) => a + l.amount_cents, 0); return { passed: e.lines.length >= 2 && d === c, detail: `debit ${d}, credit ${c}` }; });
});
Then('the entry has a number', function () {
  check(this, 'entry has a number', () => ({ passed: !!this.entry && Number.isInteger(this.entry.entry_no) && this.entry.entry_no >= 1, detail: 'entry_no ' + (this.entry && this.entry.entry_no) }));
});
Then('the entry lists {int} lines', function (n) {
  check(this, 'entry lists ' + n + ' lines', () => ({ passed: !!this.entry && this.entry.lines.length === n, detail: this.entry ? this.entry.lines.length + ' lines' : 'no entry' }));
});
Then('the entry reads back with balanced lines', function () {
  check(this, 'entry reads back balanced', () => { const e = this.api && this.api.body; if (!e || !e.lines) return { passed: false, detail: 'no body' }; const d = e.lines.filter((l) => l.side === 'debit').reduce((a, l) => a + l.amount_cents, 0); const c = e.lines.filter((l) => l.side === 'credit').reduce((a, l) => a + l.amount_cents, 0); return { passed: d === c, detail: `debit ${d}, credit ${c}` }; });
});
Then('the entry list includes the posted entry', function () {
  check(this, 'entry list includes posted', () => { const es = this.api && this.api.body && this.api.body.entries; return { passed: Array.isArray(es) && es.some((e) => e.id === this.entry.id), detail: es ? es.length + ' entries' : 'no list' }; });
});
Then('both posts are the same entry', function () {
  check(this, 'idempotent post same entry', () => { const a = this.firstEntry && this.firstEntry.body, b = this.secondEntry && this.secondEntry.body; return { passed: !!a && !!b && a.id === b.id, detail: `first ${a && a.id}, second ${b && b.id}` }; });
});

// ---- generic refusals, shared across @books @api tiers ----
Then('the request is refused with code {string}', function (code) {
  check(this, 'refused with code ' + code, () => ({ passed: !!this.api && this.api.status >= 400 && this.api.body && this.api.body.code === code, detail: this.api ? this.api.status + ' ' + j(this.api.body) : 'no response' }));
});
Then('the request is refused with status {int}', function (status) {
  check(this, 'refused with status ' + status, () => ({ passed: !!this.api && this.api.status === status, detail: this.api ? this.api.status + ' ' + j(this.api.body) : 'no response' }));
});

File: python/be/api/steps/journal_steps.py
"""Steps for be-mini-books-journal.feature. Mirror of node/be/api/steps/journal.steps.js
-- a plugin module. The generic "refused with code/status" steps live here and are
shared across every @books @api tier. The @books Before/After setup and the "store is
open" Background live in be/db/steps/books_steps.py.
"""

from __future__ import annotations

import json

from pytest_bdd import parsers, then, when

from be.api.venues.books import ApiUnreachable, MiniBooks

books = MiniBooks()


def act(qa, fn):
    if qa.source_error:
        return None
    try:
        return fn()
    except ApiUnreachable as err:
        qa.source_error = str(err)
        return None


def check(qa, description, fn):
    if qa.source_error:
        qa.unobservable(description, "the source could not be reached -- " + qa.source_error)
        return
    passed, detail = fn()
    qa.check(description, passed, detail)


def _two_line(dr, cr, amt):
    return [{"account_id": dr, "side": "debit", "amount_cents": amt}, {"account_id": cr, "side": "credit", "amount_cents": amt}]


def post_lines(qa, lines, token, date=None, key=None):
    body = {"entry_date": date or books.biz_day, "lines": lines}
    if key:
        body["idempotency_key"] = key

    def go():
        qa.api = books.post("/journal-entries", body, token=token) if token else books.post("/journal-entries", body)
        if qa.api["status"] in (201, 200):
            qa.entry = qa.api["body"]
    act(qa, go)


# ---------------------------------------------------------------- Whens


@when(parsers.re(r"^the accountant posts Dr account (?P<dr>\d+) Cr account (?P<cr>\d+) for (?P<amt>-?\d+)$"))
def posts_dr_cr(qa, dr, cr, amt):
    post_lines(qa, _two_line(int(dr), int(cr), int(amt)), qa.token)


@when(parsers.re(r"^the accountant posts Dr account (?P<dr>\d+) Cr account (?P<cr>\d+) for (?P<amt>-?\d+) dated (?P<date>\S+)$"))
def posts_dr_cr_dated(qa, dr, cr, amt, date):
    post_lines(qa, _two_line(int(dr), int(cr), int(amt)), qa.token, date=date)


@when(parsers.parse("the accountant posts a split entry debiting account {d1:d} by {a1:d} and account {d2:d} by {a2:d} crediting account {cr:d} by {ac:d}"))
def posts_split(qa, d1, a1, d2, a2, cr, ac):
    lines = [
        {"account_id": d1, "side": "debit", "amount_cents": a1},
        {"account_id": d2, "side": "debit", "amount_cents": a2},
        {"account_id": cr, "side": "credit", "amount_cents": ac},
    ]
    post_lines(qa, lines, qa.token)


@when(parsers.parse("the accountant posts an unbalanced entry of Dr {dr:d} Cr {cr:d}"))
def posts_unbalanced(qa, dr, cr):
    post_lines(qa, [{"account_id": 1, "side": "debit", "amount_cents": dr}, {"account_id": 9, "side": "credit", "amount_cents": cr}], qa.token)


@when("the accountant posts a single-line entry")
def posts_single_line(qa):
    post_lines(qa, [{"account_id": 1, "side": "debit", "amount_cents": 1000}], qa.token)


@when(parsers.re(r"^an anonymous caller posts Dr account (?P<dr>\d+) Cr account (?P<cr>\d+) for (?P<amt>-?\d+)$"))
def anon_posts(qa, dr, cr, amt):
    post_lines(qa, _two_line(int(dr), int(cr), int(amt)), None)


@when(parsers.re(r"^a forged token posts Dr account (?P<dr>\d+) Cr account (?P<cr>\d+) for (?P<amt>-?\d+)$"))
def forged_posts(qa, dr, cr, amt):
    post_lines(qa, _two_line(int(dr), int(cr), int(amt)), "usr_forged000000000000000000")


@when(parsers.re(r'^the accountant posts Dr account (?P<dr>\d+) Cr account (?P<cr>\d+) for (?P<amt>-?\d+) with key "(?P<key>[^"]+)"$'))
def posts_with_key(qa, dr, cr, amt, key):
    lines = _two_line(int(dr), int(cr), int(amt))
    post_lines(qa, lines, qa.token, key=key)
    qa.first_entry = qa.api
    qa.last_lines = lines


@when(parsers.parse('the accountant posts again with key "{key}"'))
def posts_again_with_key(qa, key):
    def go():
        qa.api = books.post("/journal-entries", {"entry_date": books.biz_day, "lines": qa.last_lines, "idempotency_key": key}, token=qa.token)
        qa.second_entry = qa.api
    act(qa, go)


@when("the accountant reads the entry")
def reads_entry(qa):
    act(qa, lambda: qa.__setattr__("api", books.get("/journal-entries/" + str(qa.entry["id"]), token=qa.token)))


@when("the accountant lists the entries")
def lists_entries(qa):
    act(qa, lambda: qa.__setattr__("api", books.get("/journal-entries", token=qa.token)))


# ---------------------------------------------------------------- Thens


@then("the entry is posted")
def entry_is_posted(qa):
    check(qa, "entry posted", lambda: (bool(qa.api) and qa.api["status"] == 201, f"{qa.api['status']} {json.dumps(qa.api['body'])}" if qa.api else "no response"))


@then("the entry is refused")
def entry_is_refused(qa):
    check(qa, "entry refused", lambda: (bool(qa.api) and qa.api["status"] >= 400, f"{qa.api['status']} {json.dumps(qa.api['body'])}" if qa.api else "no response"))


@then("the entry response balances")
def entry_response_balances(qa):
    def ev():
        e = qa.entry
        if not e or not e.get("lines"):
            return (False, f"{qa.api['status']} {json.dumps(qa.api['body'])}" if qa.api else "no entry")
        d = sum(l["amount_cents"] for l in e["lines"] if l["side"] == "debit")
        c = sum(l["amount_cents"] for l in e["lines"] if l["side"] == "credit")
        return (len(e["lines"]) >= 2 and d == c, f"debit {d}, credit {c}")
    check(qa, "entry response balances", ev)


@then("the entry has a number")
def entry_has_number(qa):
    check(qa, "entry has a number", lambda: (bool(qa.entry) and isinstance(qa.entry.get("entry_no"), int) and qa.entry["entry_no"] >= 1, "entry_no " + str(qa.entry.get("entry_no") if qa.entry else None)))


@then(parsers.parse("the entry lists {n:d} lines"))
def entry_lists_n_lines(qa, n):
    check(qa, "entry lists " + str(n) + " lines", lambda: (bool(qa.entry) and len(qa.entry["lines"]) == n, (str(len(qa.entry["lines"])) + " lines") if qa.entry else "no entry"))


@then("the entry reads back with balanced lines")
def entry_reads_back(qa):
    def ev():
        e = qa.api and qa.api["body"]
        if not e or not e.get("lines"):
            return (False, "no body")
        d = sum(l["amount_cents"] for l in e["lines"] if l["side"] == "debit")
        c = sum(l["amount_cents"] for l in e["lines"] if l["side"] == "credit")
        return (d == c, f"debit {d}, credit {c}")
    check(qa, "entry reads back balanced", ev)


@then("the entry list includes the posted entry")
def entry_list_includes(qa):
    def ev():
        es = qa.api and qa.api["body"] and qa.api["body"].get("entries")
        return (isinstance(es, list) and any(e["id"] == qa.entry["id"] for e in es), (str(len(es)) + " entries") if es else "no list")
    check(qa, "entry list includes posted", ev)


@then("both posts are the same entry")
def both_posts_same(qa):
    def ev():
        a = qa.first_entry and qa.first_entry["body"]
        b = qa.second_entry and qa.second_entry["body"]
        return (bool(a) and bool(b) and a["id"] == b["id"], f"first {a['id'] if a else None}, second {b['id'] if b else None}")
    check(qa, "idempotent post same entry", ev)


# ---- generic refusals, shared across @books @api tiers ----


@then(parsers.parse('the request is refused with code "{code}"'))
def refused_with_code(qa, code):
    def ev():
        ok = bool(qa.api) and qa.api["status"] >= 400 and qa.api["body"] and qa.api["body"].get("code") == code
        return (ok, f"{qa.api['status']} {json.dumps(qa.api['body'])}" if qa.api else "no response")
    check(qa, "refused with code " + code, ev)


@then(parsers.parse("the request is refused with status {status:d}"))
def refused_with_status(qa, status):
    check(qa, "refused with status " + str(status), lambda: (bool(qa.api) and qa.api["status"] == status, f"{qa.api['status']} {json.dumps(qa.api['body'])}" if qa.api else "no response"))
